test/hotspot/jtreg/runtime/os/TestTransparentHugePageUsage.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) Ampere Computing and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test TestTransparentHugePageUsage
+ * @bug 8315923
+ * @requires vm.gc.Parallel & os.family == "linux" & os.maxMemory > 30G
+ * @summary Check if the usage of THP is zero when enabled.
+ * @comment The test is not ParallelGC-specific, but a multi-threaded GC is
+ *          required. So ParallelGC is used here.
+ *
+ * @run main/othervm -XX:+UseTransparentHugePages
+ *                   -XX:+UseParallelGC -XX:ParallelGCThreads=${os.processors}
+ *                   -Xlog:startuptime,pagesize,gc+heap=debug
+ *                   -Xms24G -Xmx24G -XX:+AlwaysPreTouch
+ *                   runtime.os.TestTransparentHugePageUsage
+ */
```

Comment:
Requiring the test to need 24G is a lot... why does it need to be so large?

What does the test test? The problem was that pre-touching the memory would allocate small pages, and then later khugepaged would fold them into large pages at its own leisure. Your patch prevents that, so now huge pages form faster? So, the success of the patch can be described by timing?

Reply:
I changed the size to 1G which means two thp on aarch64 with 64KB page sizes. The test now checks the usage of thp, and will fail when the policy is madvise and the method of pretouch is atomic-add on aarch64. Timing is a factor, but have to be checked manually from the log, as there is no standard on how fast it should be.

Reply:
Ah, ok, now I get it. You use the size to pick out the heap from the mappings. The problem here is that 1G is a really common size, and you may e.g. find the class space first, which is 1G in default size.

May I suggest two alternatives?


Alternative 1:

call the JVM with "-Xlog:pagesize". That will print out the heap location too, at least for G1, the default allocator:

```
thomas@starfish:/shared/projects/openjdk/jdk-jdk/output-fastdebug$ ./images/jdk/bin/java -Xlog:pagesize                                   
...
[0.012s][info][pagesize] Heap:  min=8M max=16064M base=0x0000000414000000 size=16064M page_size=4K 
```

You can then parse the heap base from the output and use that to find the mapping.

Alternative 2:

Pass a really crooked size to -Xmx. Not as safe as alternative 1, but should work reasonably well.

Reply:
If you use Alternativre 1, you also can scan for UseTransparentHugePages=1 in the output and replace the mxBean call with that:

```
[0.001s][info][pagesize] UseLargePages=1, UseTransparentHugePages=1
```